Sources/VLCTranscoder.m

```diff
@@ -0,0 +1,142 @@
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
+ *****************************************************************************/
+#import "VLCTranscoder.h"
+
+#import "VLCEventManager.h"
+#import "VLCLibrary.h"
+#import "VLCLibVLCBridging.h"
+
+#include <vlc/vlc.h>
+
+@interface VLCTranscoder()
+{
+    libvlc_media_player_t *_p_mp; //player instance used for transcoding
+    dispatch_queue_t _libVLCTranscoderQueue;
+}
+- (void)registerObserversForMuxWithPlayer:(libvlc_media_player_t *)player;
+- (void)unregisterObserversForMuxWithPlayer:(libvlc_media_player_t *)player;
+@end
+
+@implementation VLCTranscoder
+
+- (instancetype)init
+{
+    if (self = [super init]) {
+        _libVLCTranscoderQueue = dispatch_queue_create("libVLCTranscoderQueue", DISPATCH_QUEUE_SERIAL);
+    }
+    return self;
+}
+
+- (BOOL)muxSubtitleFile:(NSString *)srtPath toMp4File:(NSString *)mp4Path outputPath:(NSString *)outputPath
+{
+    libvlc_media_t* p_media = libvlc_media_new_path([[VLCLibrary sharedLibrary] instance], [mp4Path UTF8String]);
+    if (p_media == NULL) {
+        NSAssert(0, @"p_media wasn't allocated");
+        return NO;
+    }
+    NSString *transcodingOptions = [NSString stringWithFormat:@":sout=#transcode{venc={module=avcodec{codec=h264_videotoolbox}, vcodec=h264},venc={module=vpx{quality-mode=2},vcodec=VP80},acodec=mpga,ab=128,channels=2,samplerate=44100,soverlay}:file{dst='%@',mux=mkv}", outputPath];
+    libvlc_media_add_option(p_media, [[NSString stringWithFormat:@"--sub-file=%@", srtPath] UTF8String]);
+    libvlc_media_add_option(p_media, [transcodingOptions UTF8String]);
+
+    _p_mp = libvlc_media_player_new_from_media( p_media );
+    if (_p_mp == NULL) {
+        NSAssert(0, @"_p_mp wasn't allocated");
+        return NO;
+    }
+
+    [self registerObserversForMuxWithPlayer:_p_mp];
+    BOOL canPlay = libvlc_media_player_play( _p_mp ) == 0;
+    if (!canPlay) {
+        NSAssert(0, @"playback failed");
+        [self unregisterObserversForMuxWithPlayer:_p_mp];
+        return NO;
+    }
+    return canPlay;
+}
+
+- (void)registerObserversForMuxWithPlayer:(libvlc_media_player_t *)player
+{
+    __block libvlc_event_manager_t * p_em = libvlc_media_player_event_manager(player);
+    if (!p_em)
+        return;
+    dispatch_sync(_libVLCTranscoderQueue,^{
```

Comment:
Why dispatch to the libVLCTranscoderQueue here but not when unregistering? Shouldn't just for safety self be retained here and released again only when unregistering to prevent that self is deallocated before having released the observers?

Reply:
Ohhhh we might have the same issue on the player :D

Reply:
how would you retain self in this case?

Reply:
```
void *retainedSelf = (__bridge_retained void *)self
```
and back

```
void *releaseSelf = (__bridge_transfer void *)self;
```
or `CFBridgingRetain` and `CFBridgingRelease`.

That said, if the lifetime is clear and this can not happen anyway, thats not needed.

Reply:
I actually think you can get i the state where the object gets destroyed while transcoding

Reply:
but that can be avoided by calling mediaPlayerStateChangeForMux on deallocation